### ISSUE TITLE
Overall code structure

### ISSUE THREAD

**Issue by frankhale** [vsClojure/vsClojure#115]
I'm looking at these issues and I'm thinking if we are going to reign these in we're going to need to tighten up the ship so to speak as far as the current code base goes. 

My initial thoughts would be to start making a clean separation of the components, for instance ClojureExtension.Parsing represents a parser for Clojure but in no way cares about or needs to care about anything related to Visual Studio integration and thus should be it's own library, completely separate from this code base. vsClojure should just be a consumer of this library. That also got me to thinking holy cow, why would we need to write a parser for Clojure when that has already been done by the compiler, is there anything in the compiler source (maybe not) that can be leveraged so that we don't have to maintain that piece ourselves? This is a big piece and will be crucial for better paren matching, paredit just to name a few.

I think ultimately there needs to be a clear delineation so that it's easy to tell where VS specific code starts and where more generic code ends. Right now I kind of see a lumping of it all together in one place and that I believe is going to make things harder to maintain. 

I think there is code within the REPL project that could itself be separate from this project. Perhaps it'd be better to have a WPF based REPL and then have extension points drawn in the sand that the VS API's could then take advantage of for integrating into VS. This REPL proper could be a separate project that doesn't live inside the vsClojure code, only the bits necessary to hook it into VS would be in there.

Hanging off that same thought there may be ways to make most of the project management generic so that the VS parts can be abstracted away leaving just a clean project management layer that VS hooks into.

I guess what I'm ultimately saying is that we probably shouldn't dump all this code together in one place making it tightly coupled.

Perhaps we can start a new project that builds up vsClojure from scratch? If anything it'd be a great way to learn of all the various components necessary and start working this clean separation of concerns in.

**Comment by frankhale** [comment]
Of course I could be way off base!

**Comment by speige** [comment]
I'm in agreement with most of what you're saying. The tight coupling would make it hard for anyone to take pieces & integrate them into other software, like sharpdevelop.

I'm not sure fixing the coupling will make bugs easier to fix, but it'd definitely be cleaner & more reusable code.

the compiler does have a lot of things built in that we can use. I've tried to use that in metadata.clj. We should probably try to convert some of the parser c# library into clojure code. The only downside to eating our own dog food is that clojureClr takes a long time to initialize. I've tried to work around that by starting it up in a separate thread on visual studio loading. However, by using our own product, we can see what the pain points are for actual developers using clojure & the extension & we can get them fixed.

the current repl has hardly any code because its essentially a wrapper around clojure.main.exe, I'm not sure if it'd be useful to anyone.

I think before we consider splitting anything off into a separate repo, we should decouple into separate .csproj within the same solution (& maybe combine some). Once we've cleaned up the current repo, we can discuss whether any part would be useful to others in a separate repo.

I'm not sure what you mean by the project management stuff.

I haven't gone through any visual studio extension tutorials. I've just been adding features & fixing bugs in the existing code by Jon. Its been several years since he started it, so there's probably some new templates that we could use. If we migrate to a new one by copying/pasting code from the existing solution into a new one, it'd help clean things up & possibly fix a few bugs.

**Comment by frankhale** [comment]
My overall thought was to separate reusable code that doesn't depend on VS. So that said, the REPL is pretty much a reusable thing that doesn't necessarily depend on VS but can be integrated into it. This is probably a poor example of what I mean. What is better is some of the code within the Utilities project, for instance the code within the IO folder. This code is unlikely to change, thus, could be made into a separate library and simply used within the extension. This could reduce the overall footprint of the code we have to maintain. I'm sure there is some other stuff in there that could be moved into it's own library that lives outside the extension code. This would be purely to reduce the complexity of what we have to deal with.

**Comment by frankhale** [comment]
```
I'm not sure what you mean by the project management stuff.
```

I was referring to ClojureExtension.Project, perhaps there are bits in there that can be made more generic and abstracted away because eventually it'd be nice to have some sort of bridge between Leiningen and Visual Studio so that projects could be managed in a more generic way. Now that sounds all fanciful but I'm not quite sure yet how that would all work out. It seems at least by glancing through the ClojureExtension.Project that it's pretty tightly coupled and assumes Visual Studio project management and that's it.

**Comment by speige** [comment]
Unless we believe someone else will need the code for another project, I don't see the value in moving it to a separate repo. If we believe the code won't be changed in the future, then it shouldn't cause any more maintenance being inside the solution compared to being in a separate repo. However, if there will be changes in the future, being in another repo will increase maintenance costs slightly.

I can definitely see some of the features being useful to other projects. I think we should work on re-organizing the code within the solution & once we're happy with it, decide which pieces should be moved to a separate repo.

**Comment by frankhale** [comment]
Sounds good!

**Comment by frankhale** [comment]
I think maybe the best course of action is to start over with a clean repo and start building vsClojure up from the ground. I've been reading a lot of the documentation on Visual Studio extensibility and it's starting to make some sense now. I was thinking of two areas initially to start on and that is a better UI for the REPL and language provider that has good syntax highlighting that is customizable and rainbow parenthesis for starters.

**Comment by speige** [comment]
I'm fine with this. I assume there's a new template for visual studio extensibility we could use which would fix a lot of our existing issues.

Before we start, are there architectural decisions we need to make? For Example: What Visual Studio versions will the extension support? Do we want to code everything in Clojure or C# or a mix? How will we decouple visual studio glue from core code? Do we want to maintain a suite of unit or integration tests (If Unit, do we need to use a Dependency Injection framework to reduce coupling)? Do we want to follow a certain coding standard?

If we're starting from scratch, do we want to reconsider the license & copyright & if so, do we need to be careful to not take code from the original repo & rewrite features from scratch?

Where do you want to store this new solution? Do we want a brand new repo, or should we just create a fork, create a new solution in the fork, move the old code into the new solution (cleaning up & rearranging as we go) & then pull back into main?

**Comment by frankhale** [comment]
```
I'm fine with this. I assume there's a new template for visual studio extensibility we could use which would fix a lot of our existing issues.
```

I'm sure it's been updated for the latest version of Visual Studio but I don't actually have any references to prove that. 

```
Before we start, are there architectural decisions we need to make? For Example: What Visual Studio versions will the extension support?
```

I think it'd be nice to support at least 2012 and 2013 (those being the easiest because there is likely not much differences between the two). 2010 support would be a nice to have. That said, I'm so new to Visual Studio extensibility that I don't know what it takes to support older versions of VS. 

```
Do we want to code everything in Clojure or C# or a mix?
```

Writing in Clojure seems extremely interesting but I don't know how much difficulty we'd run into as far as the templates go and getting the environment working with a proper debug / run cycle using the experimental instance of VS. That said, I think we should pick one or the other, I think mixing them makes the maintenance aspects more difficult.

```
How will we decouple visual studio glue from core code?
```

One area that I think of instantly is the Clojure parser and code surrounding that which will be used by the Language Service. A lot of that code to discover information about the underlying source file will be pretty generic and be able to easily become part of a separate library that can be used by the Language Service. I'm also fairly certain that the REPL window can be abstracted into it's own WPF library without any VS SDK code living inside of it. We'll just define a nice API in order to interact with it as well as be able to host it inside of VS. 

The part that is less clear to me is the project management piece. Part of me wants this to be somewhat generic in that it can work either in a VS context or as part of a Leiningen context. That said, I don't know how that will be worked yet because I am not familiar enough at this point with how the current project management works.

The following talk really drove home that the extension doesn't have to be an all or nothing proposition and that all of these pieces can be built up over time while still providing functionality incrementally. This particular talk discusses providing a language service.

http://channel9.msdn.com/blogs/vsipmarketing/vsx212-adding-a-language-service-into-visual-studio-2010

```
 Do we want to maintain a suite of unit or integration tests (If Unit, do we need to use a Dependency Injection framework to reduce coupling)? Do we want to follow a certain coding standard?
```

I think at the very least a suite of unit tests, I think integration tests would be a nice to have. However, we can do both if that is the consensus. I think maybe it'll be easier to just start with unit tests at first.

DI might be a bridge too far, I guess it all depends on how we think of the SDK glue part of the project. I suppose using a DI approach would give us an easier road on reusing the part of the code that actually touches VS. I'd opt for not using this approach initially. I'd just exercise caution in introducing too many layers of complexity right off the bat. I think once the ball gets rolling and some momentum gets built up it'll be easier to absorb a layer of complexity like this.

```
Do we want to follow a certain coding standard?
```

I'm not sure I'm tracking on this question, is this in reference to the unit tests and such? Design methodology? TDD, BDD, stuff like that?

```
If we're starting from scratch, do we want to reconsider the license & copyright & if so, do we need to be careful to not take code from the original repo & rewrite features from scratch?
```

Yeah, I'd reconsider the license to the extent that _I_ prefer a really liberal license. Perhaps MIT provides this, I'm not hugely familiar with it. Your thoughts on this would be great, I'm not particular married to the ideals set forth between various licenses, my only thing is that I like the developers utilizing the code to have the maximum ability to use it without restrictions. The Clojure world tends to prefer the Eclipse Public License. Any incite you can provide would be valuable on this front.

It's easy to stand up and cry "REWRITE" but my overall thought was that diving into the existing code may lead to confusion (on my part) in the long run because it's already implemented and I think  it may be difficult to diagnose if it's done right or wrong. Given the video that I posted what I took away from watching that is that these pieces could be pieced together and still provide a workable solution so that starting over with just one piece and then adding to it would be much easier in the long run to understand. A huge part of this work is simply climbing the Visual Studio extension mountain and I don't think it can be done nearly as easily when diving straight into an already implemented project. My brain works best when I can start and build upon something piece by piece.

As far as new code or reusing code that all depends. If there is value in reusing the existing code then lets do it. I'd say probably the only area I'd reuse right now is probably the Clojure parser but then again I don't know (haven't investigated yet) if this parser is completely capable of doing all that we'd want to be able to do as far as the Language Service goes. I'd like to be able ultimately to provide some seriously kick ass syntax highlighting, intellisense, rainbow parenthesis, code formatting, outlining (I'd like some sort of capability for a feature like C# regions). Something like paredit and error highlighting.

```
Where do you want to store this new solution? Do we want a brand new repo, or should we just create a fork, create a new solution in the fork, move the old code into the new solution (cleaning up & rearranging as we go) & then pull back into main?
```

This I'm not sure of yet. If there is interest in rewriting would we keep the same name vsClojure or come up with a new name? It may be time to reconsider a new name if it's all new work. I'd think it'd be a clean repo rather than a fork.

**Comment by frankhale** [comment]
So that said, I think the first area of work should be the Clojure parser and the Language Service. The first milestone would be to get syntax coloring going for Clojure / Clojurescript files.

Another piece that would be relatively easy to get started on is a really kick ass WPF based Clojure REPL and this could be all wrapped up in a library that would provide some nice API's so that VS could host the window inside a tool window as well as interact with it (eg. send code to it either by line or by file, etc..).

**Comment by speige** [comment]
My preference would be to write it 100% in Clojure. I'm certain we'll have pain points with that (Debugging, COMInterop, Attribute annotation on classes). However, If we don't eat our own dog food, how will we know if our extension is actually useful to a Clojure/VisualStudio developer? Also, as much as the ClojureCLR community needs a good tool for developing Clojure, it also needs a good Clojure runtime. We might come up with a great extension, without helping fix any of the pain points in the runtime itself. If we're not using the Clojure runtime to build our extension, we won't find the pain points in the runtime. If the community doesn't have both a good runtime & a good tool, neither is useful.

Also, I'm pretty busy so it's hard to find time to work on side projects. However, if I'm going to write code on the side, I'll be a lot more excited to write it if it's Clojure.

Shall we vote on a name & start the new repo?

Hopefully we can get the new repo up to equivalent functionality quickly, but we should try to at least provide minimal support for vsClojure in the interim.

By coding standard, I guess I was mostly talking about formatting. So, if we write the extension in Clojure maybe we pick https://github.com/bbatsov/clojure-style-guide for C# maybe we download a styleCop template we like. This could also include things like not using static variables (if we want to make mocking in our c# unit testing easier), or other code smells. I don't think we really need a document to list coding rules, I think we just code how we think is best & if someone has a suggestion for a change, they can speak up.

If our reason for starting from scratch is because it's difficult for a new person to understand the existing code, I don't think we have a legitimate reason for a rewrite. However, if we have ideas for a better infrastructure that has long-term benefits (ease of maintenance, etc), then maybe we have a valid reason.

**Comment by frankhale** [comment]
I like the idea of using Clojure for the rewrite and I also agree with the style guide you linked to. As far as a name and a new repo we can probably hold off on that for a little while.

I think for the time being I'm going to take the next several weeks to dissect vsClojure as well as absorb all of the reference material I can get my hands on. I think that is probably the most prudent approach for me for the near term.

**Comment by frankhale** [comment]
I'm going to close this for now. It can be reopened when the time arises.
